homeassistant/components/ecowitt/__init__.py

```diff
@@ -0,0 +1,178 @@
+"""The Ecowitt Weather Station Component."""
+from __future__ import annotations
+
+from dataclasses import dataclass
+import logging
+import time
+
+from pyecowitt import WINDCHILL_HYBRID, WINDCHILL_NEW, WINDCHILL_OLD, EcoWittListener
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import (
+    CONF_PORT,
+    CONF_UNIT_SYSTEM_IMPERIAL,
+    CONF_UNIT_SYSTEM_METRIC,
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.dispatcher import (
+    async_dispatcher_connect,
+    async_dispatcher_send,
+)
+from homeassistant.helpers.entity import DeviceInfo, Entity
+
+from .const import (
+    CONF_UNIT_BARO,
+    CONF_UNIT_LIGHTNING,
+    CONF_UNIT_RAIN,
+    CONF_UNIT_WIND,
+    CONF_UNIT_WINDCHILL,
+    DATA_MODEL,
+    DATA_PASSKEY,
+    DATA_STATIONTYPE,
+    DOMAIN,
+    ECOWITT_PLATFORMS,
+    SIGNAL_NEW_SENSOR,
+    SIGNAL_REMOVE_ENTITIES,
+    SIGNAL_UPDATE,
+    W_TYPE_HYBRID,
+    W_TYPE_NEW,
+    W_TYPE_OLD,
+)
+
+NOTIFICATION_ID = DOMAIN
+NOTIFICATION_TITLE = "Ecowitt config migrated"
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@dataclass
+class EcowittData:
+    """Internal data class for the ecowitt integration."""
+
+    client: type[EcoWittListener]
+    registered_devices: list[str]
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up the Ecowitt component from UI."""
+
+    if not entry.options:
+        options = {
+            CONF_UNIT_BARO: CONF_UNIT_SYSTEM_METRIC,
+            CONF_UNIT_WIND: CONF_UNIT_SYSTEM_IMPERIAL,
+            CONF_UNIT_RAIN: CONF_UNIT_SYSTEM_IMPERIAL,
+            CONF_UNIT_LIGHTNING: CONF_UNIT_SYSTEM_IMPERIAL,
+            CONF_UNIT_WINDCHILL: W_TYPE_HYBRID,
+        }
+        hass.config_entries.async_update_entry(entry, options=options)
+
+    # setup the base connection
+    data = EcowittData(EcoWittListener(port=entry.data[CONF_PORT]), [])
+    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = data
```

Comment:
I would split this into two lines.

Reply:
do you mean the hass.data.setdefault?  that seems to be the convention now.  I think in balboa I had it split and then was told to merge it like this.

Reply:
oh ok.

Reply:
For reference: https://github.com/home-assistant/core/pull/59234#discussion_r748936790
I think it was mostly to drop the `if` statement, but if it's ok for `bdraco` it's ok for me.